core/src/main/java/org/elasticsearch/search/builder/NewSearchSourceBuilder.java

```diff
@@ -0,0 +1,1077 @@
+import org.elasticsearch.search.suggest.SuggestBuilder;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * A search source builder allowing to easily build search source. Simple
+ * construction using
+ * {@link org.elasticsearch.search.builder.NewSearchSourceBuilder#searchSource()}.
+ *
+ * @see org.elasticsearch.action.search.SearchRequest#source(NewSearchSourceBuilder)
+ */
+public class NewSearchSourceBuilder extends ToXContentToBytes implements Writeable<NewSearchSourceBuilder> {
+
+    public static final ParseField FROM_FIELD = new ParseField("from");
+    public static final ParseField SIZE_FIELD = new ParseField("size");
+    public static final ParseField TIMEOUT_FIELD = new ParseField("timeout");
+    public static final ParseField TERMINATE_AFTER_FIELD = new ParseField("terminate_after");
+    public static final ParseField QUERY_FIELD = new ParseField("query");
+    public static final ParseField POST_FILTER_FIELD = new ParseField("post_filter");
+    public static final ParseField MIN_SCORE_FIELD = new ParseField("min_score");
+    public static final ParseField VERSION_FIELD = new ParseField("version");
+    public static final ParseField EXPLAIN_FIELD = new ParseField("explain");
+    public static final ParseField _SOURCE_FIELD = new ParseField("_source");
+    public static final ParseField FIELDS_FIELD = new ParseField("fields");
+    public static final ParseField FIELDDATA_FIELDS_FIELD = new ParseField("fielddata_fields");
+    public static final ParseField SCRIPT_FIELDS_FIELD = new ParseField("script_fields");
+    public static final ParseField SCRIPT_FIELD = new ParseField("script");
+    public static final ParseField SORT_FIELD = new ParseField("sort");
+    public static final ParseField TRACK_SCORES_FIELD = new ParseField("track_scores");
+    public static final ParseField INDICES_BOOST_FIELD = new ParseField("indices_boost");
+    public static final ParseField AGGREGATIONS_FIELD = new ParseField("aggregations");
+    public static final ParseField HIGHLIGHT_FIELD = new ParseField("highlight");
+    public static final ParseField INNER_HITS_FIELD = new ParseField("inner_hits");
+    public static final ParseField SUGGEST_FIELD = new ParseField("suggest");
+    public static final ParseField RESCORE_FIELD = new ParseField("rescore");
+    public static final ParseField STATS_FIELD = new ParseField("stats");
+
+    /**
+     * A static factory method to construct a new search source.
+     */
+    public static NewSearchSourceBuilder searchSource() {
+        return new NewSearchSourceBuilder();
+    }
+
+    /**
+     * A static factory method to construct new search highlights.
+     */
+    public static HighlightBuilder highlight() {
+        return new HighlightBuilder();
+    }
+
+    private QueryBuilder<?> queryBuilder;
+
+    private QueryBuilder<?> postQueryBuilder;
```

Comment:
given that we have the `queryBuilder` now we could drop post after we get this in right? Can we add a TODO that we fix this afterwards?